# RNASeq pipeline. From FASTQ to DEGs and GO analysis.
This repository contains all the code used in the analysis of RNA-seq data from rice (_O. sativa_). You can see two different directories:

- **Pre-processing, Alignment & Quantification.** Bash scripts corresponding to the following stages are deposited here: pre-processing (*RawData_Fastqc.sh* & *TrimmedData_Fastqc.sh*), alignment (*Alignment_hisat2.sh*), assembly and quantification (*Assembly_&_Quantification_stringtie.sh*).
  
- **Differential Expression Analysis & GO Enrichment.** R scripts to perform all the steps necessary to analyze a gene expression matrix from RNA sequencing using the new Tuxedo protocol.


Please read all the comments in the scripts carefully. If you have any questions, please do not hesitate to contact me at the follow e-mail address:

**Víctor J. Fernández Ramírez** 

*victor.fernandez@imibic.org* 

*Bioinformatician* 
